docs/architecture/adr-069-unordered-account.md

```diff
@@ -0,0 +1,194 @@
+
+## Decision
+
+Add an extra nonce lane to support optional un-ordered transaction inclusion, the default lane provides ordered semantic the same as before, the new one will provide un-ordered semantic. The transaction can choose which lane to use.
+
+One of the design goals is to keep minimal overhead and breakage to the existing users who don't use the new feature.
+
+### Transaction Format
+
+It doesn't change the transaction format itself, but re-use the MSB(most significant bit) of the existing 64bits nonce value to identify the lane, `0` being the default ordered lane, `1` being the new un-ordered lane.
+
+### Account State
+
+The new nonce lane needs to add some optional fields to the account state:
+
+```golang
+type UnOrderedNonce struct {
+  Sequence uint64
+  Timestamp uint64  // the block time when the Sequence is updated
+  Gaps IntSet
+}
+
+type Account struct {
+  // default to `nil`, only initialized when the new feature is first used.
+  unorderedNonce *UnOrderedNonce
+}
+```
+
+The un-ordered nonce state includes a normal sequence value plus the set of unused(gap) values in recent history, these recorded gap values can be reused by future transactions, after used they are removed from the set and can't be used again, the gap set has a maximum capacity to limit the resource usage, when the capacity is reached, the oldest gap value is removed, which also makes the pending transaction using that value as nonce will not be accepted anymore.
+
+### Expiration
+
+It would be good to expire the gap nonces after certain timeout is reached, to mitigate the risk that a middleman intercept an old transaction and re-execute it in a longer future, which might cause unexpected result.
+
+### Prototype Implementation
+
+The prototype implementation use a roaring bitmap to record these gap values.
+
+```golang
+const (
+  MSBCheckMask = 1 << 63
+  MSBClearMask = ^(1 << 63)
+)
+
+// CheckNonce switches to un-ordered lane if the MSB of the nonce is set.
+func (acct *Account) CheckNonce(nonce uint64, blockTime uint64) error {
+  if nonce & MSBCheckMask != 0 {
+    nonce &= MSBClearMask
+
+    if acct.unorderedNonce == nil {
+      acct.unorderedNonce = NewUnOrderedNonce()
+    }
+    return acct.unorderedNonce.CheckNonce(nonce, blockTime)
+  }
+
+  // current nonce logic
+}
+```
+
+
+
+```golang
+const (
+  // GapsCapacity is the capacity of the set of gap values.
+  GapsCapacity = 1024
+  // MaxGap is the maximum gaps a new nonce value can introduce
+  MaxGap = 1024
+  // GapsExpirationDuration is the duration in seconds for the gaps to expire
+  GapsExpirationDuration = 60 * 60 * 24
+)
+
+// getGaps returns the gap set, or create a new one if it's expired
+func(u *UnOrderedNonce) getGaps(blockTime uint64) *IntSet {
```

Comment:
Nitpicking, `u` is not a nonce, but a nonce manager, or nonce gap manager.

Reply:
done, changed to `UnorderedNonceManager`